resolwe_bio/tools/expression_rpkum.py

```diff
@@ -0,0 +1,99 @@
+import argparse
+import gzip
+import math
+import sys
+
+import pandas as pd
+
+from os.path import basename
+from resolwe_runtime_utils import error
+
+
+def parse_arguments():
+    """Parse command line arguments."""
+    parser = argparse.ArgumentParser(description="Normalize expressions using RPKUM.")
+    parser.add_argument('-c', '--counts', required=True, help="Raw counts file.")
+    parser.add_argument('-m', '--mappability', required=True, help="Mappability file.")
+    parser.add_argument('-o', '--output', required=True, help="Output file name.")
+    return parser.parse_args()
+
+
+def parse_expression_file(exp_file):
+    """Parse expression file to a Pandas dataframe."""
+    try:
+        df = pd.read_csv(
+            exp_file,
+            sep='\t',
+            compression='gzip',
+            usecols=['Gene', 'Expression'],
+            index_col='Gene',
+            dtype={
+                'Gene': str,
+                'Expression': float,
+            },
+        )
+        df.dropna(inplace=True)
+        return df
+    except (ValueError, OSError) as parse_error:
+        print(error("Failed to read input file {}. {}".format(basename(exp_file), parse_error)))
+        sys.exit(1)
+
+
+def parse_mapability_file(mapability_file):
+    """Parse mapability file to a Pandas dataframe."""
+    try:
+        mappability = pd.read_csv(
+            mapability_file,
+            sep='\t',
+            usecols=['gene_id', 'coverage'],
+            index_col='gene_id',
+            dtype={
+                'gene_id': str,
+                'coverage': float,
+            },
+        )
+        mappability.dropna(inplace=True)
+        return mappability
+    except (ValueError, OSError) as parse_error:
+        print(error("Failed to read mappability file {}. {}".format(basename(exp_file), parse_error)))
+        sys.exit(1)
+
+
+def get_coverage(gene_id, mappability):
+    """Return gene coverage from the mapability file."""
+    try:
+        return mappability.loc[gene_id]['coverage']
+    except KeyError:
+        print(error("Feature ID {} is not present in the mappability file. "
+                    "Make sure that the expressions and mappability file are "
+                    "derived from the same annotations (GTF/GFF) file.".format(gene_id)))
+        sys.exit(1)
+
+
+def rpkum(rc, cov, lib_size):
+    """Compute RPKUM."""
+    if cov == 0:
+        return 0
+    else:
+        return (math.pow(10, 9) * rc) / (lib_size * cov)
+
+def main():
+    """Invoke when run directly as a program."""
+    args = parse_arguments()
+
+    mappability = parse_mapability_file(args.mappability)
+    expression = parse_expression_file(args.counts)
+    lib_size = expression['Expression'].sum()
+
+    # Compute RPKUM expression
+    expression['Expression'] = expression.apply(
+        lambda row: rpkum(row.Expression, get_coverage(row.name, mappability), lib_size),
+        axis=1
+    )
+    expression.sort_index(inplace=True)
```

Comment:
Are you sure? I presume that the input expression file with raw counts is already sorted. In this case, this line doesn't change anything. If the expression file with raw counts is unsorted it makes more sense to keep the same order of genes in RPKUM expression file as well.

Otherwise do not forget to change `expression` to `df` here.